### PR TITLE
Implement issue 13433 - add option for using coarser realtime clock.

### PR DESCRIPTION
**Pull request by jmdavis**
This adds an option for getting the time using the coarser, faster
realtime clock on systems that support it for those who need to get the
time frequently (and thus quickly) but don't need it to be particularly
precise.

https://issues.dlang.org/show_bug.cgi?id=13433

The OSX change is because it's the only one using `gettimeofday`. I figured that it was better to change that to go off of the OS version rather than whether `clock_gettime` existed or not.

### CODE REVIEW

**Comment by jmdavis** [comment]
Okay. This has been updated to use `core.time.ClockType` rather than a boolean for whether the clock should be coarse or not (so, in addition to the normal clock, there is now a coarse clock, a precise clock, and a clock with second precision), and the argument is a compile-time argument, since that works better (and is what core.time is doing with `MonoTime`, though in this case, it's just a couple of functions being templated rather than a whole type).

File: std/datetime.d [hunk 1]
```diff
@@ -310,84 +310,184 @@ public:
...
-                       convert!("usecs", "hnsecs")(tv.tv_usec) +
-                       hnsecsToUnixEpoch;
-            }
-        }
+        auto norm1 = Clock.currStdTime;
+        auto norm2 = Clock.currStdTime;
+        assert(norm1 <= norm2, format("%s %s", norm1, norm2));
+        assert(abs(norm1 - norm2) <= limit);
+
+        auto coarse1 = Clock.currStdTime!(ClockType.coarse);
```
Comment (DmitryOlshansky) [inline std/datetime.d]: These 3 quartets of lines are exactly the same - foreach over a `TypeTuple!(ClockType.coarse, ClockType.precise, ClockType.second)` ?

File: std/datetime.d [hunk 2]
```diff
@@ -310,84 +310,184 @@ public:
...
+        import std.math : abs;
+        import std.format : format;
+        enum limit = convert!("seconds", "hnsecs")(2);
 
-                return convert!("seconds", "hnsecs")(tv.tv_sec) +
-                       convert!("usecs", "hnsecs")(tv.tv_usec) +
-                       hnsecsToUnixEpoch;
-            }
-        }
+        auto norm1 = Clock.currStdTime;
```
Comment (DmitryOlshansky) [inline std/datetime.d]: oh and these 4 lines as well.

File: std/datetime.d [hunk 3]
```diff
@@ -310,84 +310,184 @@ public:
...
             immutable unixTimeD = currTime().toUnixTime();
             immutable unixTimeC = core.stdc.time.time(null);
-            immutable diff = unixTimeC - unixTimeD;
-
-            assert(diff >= -2);
-            assert(diff <= 2);
+            assert(std.math.abs(unixTimeC - unixTimeD) <= 2);
         }
+
+        auto norm1 = Clock.currTime;
```
Comment (DmitryOlshansky) [inline std/datetime.d]: I think I've seen this exact code elsewhere. I believe that unittests are as good place for using a bit of DRY as anything else.

**Comment by jmdavis** [comment]
Okay. The tests were updated to use `TypeTuple` as suggested.

**Comment by d-random-contributor** [comment]
FWIW, C `time` function returns a value with an implementation-defined meaning, it's neither unix time, nor UTC, nor a number of seconds. It's just unix happens to define C time to be unix time. You should use `gmtime` or `localtime` to get anything meaningful from it.

File: std/datetime.d [hunk 4]
```diff
@@ -310,84 +310,174 @@ public:
...
+                else static if(clockType == ClockType.precise) alias clockArg = CLOCK_REALTIME_PRECISE;
+                else static if(clockType == ClockType.second)  alias clockArg = CLOCK_SECOND;
+                else static assert(0, "Previous static if is wrong.");
                 timespec ts;
-
-                if(clock_gettime(CLOCK_REALTIME, &ts) != 0)
-                    throw new TimeException("Failed in clock_gettime().");
-
+                if(clock_gettime(clockArg, &ts) != 0)
+                    throw new TimeException("Call to gettimeofday() failed");
```
Comment (schveiguy) [inline std/datetime.d]: Shouldn't this say "Call to clock_gettime"?

Comment (jmdavis) [inline std/datetime.d]: Yeah. It's `gettimeofday` for Mac OS X, and `clock_gettime` for Linux and FreeBSD, and apparently, I screwed up the message when I was doing some copy-pasting. Fixed now.

**Comment by jmdavis** [comment]
Any more comments on this or things that need to be changed before this can be merged?

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Pull request merged**